File: internal/telemetry/provider.go
package telemetry

import (
	"context"
	"fmt"
	"regexp"
	"time"

	"github.com/sirupsen/logrus"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/metric/noop"
	"go.opentelemetry.io/otel/propagation"
	metricSdk "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	traceSdk "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.20.0"
	"go.opentelemetry.io/otel/trace"

	"goauthentik.io/internal/config"
)

// Provider manages OpenTelemetry telemetry configuration for Go services
type Provider struct {
	config         *config.Config
	logger         *logrus.Logger
	initialized    bool
	tracerProvider *traceSdk.TracerProvider
	meterProvider  *metricSdk.MeterProvider
	sampler        *AdaptiveSampler

	// Metrics
	requestCounter  metric.Int64Counter
	requestDuration metric.Float64Histogram
	authCounter     metric.Int64Counter
	ldapOpCounter   metric.Int64Counter
	ldapOpDuration  metric.Float64Histogram
}

// AdaptiveSampler provides intelligent sampling for Go services
type AdaptiveSampler struct {
	defaultRate         float64
	errorRate           float64
	healthCheckPatterns []*regexp.Regexp
}

// NewAdaptiveSampler creates a new adaptive sampler
func NewAdaptiveSampler(defaultRate, errorRate float64) *AdaptiveSampler {
	return &AdaptiveSampler{
		defaultRate: defaultRate,
		errorRate:   errorRate,
		healthCheckPatterns: []*regexp.Regexp{
			regexp.MustCompile(`/-/health/?$`),
			regexp.MustCompile(`/-/metrics/?$`),
			regexp.MustCompile(`/-/ready/?$`),
			regexp.MustCompile(`/-/live/?$`),
		},
	}
}

// ShouldSample implements custom sampling logic
func (s *AdaptiveSampler) ShouldSample(parameters traceSdk.SamplingParameters) traceSdk.SamplingResult {
	// Extract URL path from span attributes
	var urlPath string
	for _, attr := range parameters.Attributes {
		if attr.Key == "http.target" {
			urlPath = attr.Value.AsString()
			break
		}
	}

	// Never sample health check endpoints
	for _, pattern := range s.healthCheckPatterns {
		if pattern.MatchString(urlPath) {
			return traceSdk.SamplingResult{
				Decision: traceSdk.Drop,
			}
		}
	}

	// For now, sample everything else (can be enhanced with error detection)
	return traceSdk.SamplingResult{
		Decision: traceSdk.RecordAndSample,
	}
}

// Description returns sampler description
func (s *AdaptiveSampler) Description() string {
	return "AdaptiveSampler"
}

// NewProvider creates a new telemetry provider
func NewProvider(cfg *config.Config, logger *logrus.Logger) *Provider {
	return &Provider{
		config:  cfg,
		logger:  logger,
		sampler: NewAdaptiveSampler(cfg.Telemetry.OTLP.TracesSampleRate, 1.0),
	}
}

// IsEnabled checks if telemetry is enabled
func (p *Provider) IsEnabled() bool {
	return p.config.Telemetry.OTLP.Enabled
}

// Initialize sets up OpenTelemetry with OTLP configuration
func (p *Provider) Initialize(ctx context.Context, serviceName string) error {
	if p.initialized {
		return nil
	}

	if !p.IsEnabled() {
		p.logger.Debug("OTLP telemetry disabled")
		return nil
	}

	// Create resource with service attributes
	res, err := p.createResource(serviceName)
	if err != nil {
		return fmt.Errorf("failed to create resource: %w", err)
	}

	// Initialize tracing
	if err := p.setupTracing(ctx, res); err != nil {
		return fmt.Errorf("failed to setup tracing: %w", err)
	}

	// Initialize metrics
	if err := p.setupMetrics(ctx, res); err != nil {
		return fmt.Errorf("failed to setup metrics: %w", err)
	}

	// Set global propagator
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	// Create application metrics
	if err := p.createMetrics(); err != nil {
		return fmt.Errorf("failed to create metrics: %w", err)
	}

	p.initialized = true
	p.logger.Info("OpenTelemetry telemetry initialized successfully")
	return nil
}

// createResource creates OpenTelemetry resource with service attributes
func (p *Provider) createResource(serviceName string) (*resource.Resource, error) {
	attrs := []attribute.KeyValue{
		semconv.ServiceName(p.getServiceName(serviceName)),
		semconv.ServiceVersion(p.getServiceVersion()),
	}

	// Add custom resource attributes
	for key, value := range p.config.Telemetry.OTLP.ResourceAttributes {
		attrs = append(attrs, attribute.String(key, value))
	}

	return resource.NewWithAttributes(
		semconv.SchemaURL,
		attrs...,
	), nil
}

// setupTracing configures OpenTelemetry tracing
func (p *Provider) setupTracing(ctx context.Context, res *resource.Resource) error {
	endpoint := p.config.Telemetry.OTLP.Endpoint
	if endpoint == "" {
		return nil
	}

	var exporter traceSdk.SpanExporter
	var err error

	// Create appropriate exporter based on protocol
	if p.config.Telemetry.OTLP.Protocol == "http" {
		opts := []otlptracehttp.Option{
			otlptracehttp.WithEndpoint(endpoint + "/v1/traces"),
		}
		if len(p.config.Telemetry.OTLP.Headers) > 0 {
			opts = append(opts, otlptracehttp.WithHeaders(p.config.Telemetry.OTLP.Headers))
		}
		exporter, err = otlptracehttp.New(ctx, opts...)
	} else {
		// Default to gRPC
		opts := []otlptracegrpc.Option{
			otlptracegrpc.WithEndpoint(endpoint),
		}
		if len(p.config.Telemetry.OTLP.Headers) > 0 {
			opts = append(opts, otlptracegrpc.WithHeaders(p.config.Telemetry.OTLP.Headers))
		}
		exporter, err = otlptracegrpc.New(ctx, opts...)
	}

	if err != nil {
		return fmt.Errorf("failed to create trace exporter: %w", err)
	}

	// Create tracer provider with custom sampler
	p.tracerProvider = traceSdk.NewTracerProvider(
		traceSdk.WithBatcher(exporter),
		traceSdk.WithResource(res),
		traceSdk.WithSampler(p.sampler),
	)

	// Set global tracer provider
	otel.SetTracerProvider(p.tracerProvider)

	return nil
}

// setupMetrics configures OpenTelemetry metrics
func (p *Provider) setupMetrics(ctx context.Context, res *resource.Resource) error {
	endpoint := p.config.Telemetry.OTLP.Endpoint
	if endpoint == "" {
		return nil
	}

	var exporter metricSdk.Exporter
	var err error

	// Create appropriate exporter based on protocol
	if p.config.Telemetry.OTLP.Protocol == "http" {
		opts := []otlpmetrichttp.Option{
			otlpmetrichttp.WithEndpoint(endpoint + "/v1/metrics"),
		}
		if len(p.config.Telemetry.OTLP.Headers) > 0 {
			opts = append(opts, otlpmetrichttp.WithHeaders(p.config.Telemetry.OTLP.Headers))
		}
		exporter, err = otlpmetrichttp.New(ctx, opts...)
	} else {
		// Default to gRPC
		opts := []otlpmetricgrpc.Option{
			otlpmetricgrpc.WithEndpoint(endpoint),
		}
		if len(p.config.Telemetry.OTLP.Headers) > 0 {
			opts = append(opts, otlpmetricgrpc.WithHeaders(p.config.Telemetry.OTLP.Headers))
		}
		exporter, err = otlpmetricgrpc.New(ctx, opts...)
	}

	if err != nil {
		return fmt.Errorf("failed to create metrics exporter: %w", err)
	}

	// Create meter provider
	p.meterProvider = metricSdk.NewMeterProvider(
		metricSdk.WithResource(res),
		metricSdk.WithReader(metricSdk.NewPeriodicReader(exporter,
			metricSdk.WithInterval(60*time.Second),
		)),
	)

	// Set global meter provider
	otel.SetMeterProvider(p.meterProvider)

	return nil
}

// createMetrics creates application-specific metrics
func (p *Provider) createMetrics() error {
	meter := otel.Meter("authentik.outpost")

	var err error

	// HTTP request metrics
	p.requestCounter, err = meter.Int64Counter(
		"authentik_requests_total",
		metric.WithDescription("Total number of HTTP requests"),
	)
	if err != nil {
		return fmt.Errorf("failed to create request counter: %w", err)
	}

	p.requestDuration, err = meter.Float64Histogram(
		"authentik_request_duration_seconds",
		metric.WithDescription("Duration of HTTP requests"),
		metric.WithUnit("s"),
	)
	if err != nil {
		return fmt.Errorf("failed to create request duration histogram: %w", err)
	}

	// Authentication metrics
	p.authCounter, err = meter.Int64Counter(
		"authentik_authentication_total",
		metric.WithDescription("Total number of authentication attempts"),
	)
	if err != nil {
		return fmt.Errorf("failed to create auth counter: %w", err)
	}

	// LDAP-specific metrics
	p.ldapOpCounter, err = meter.Int64Counter(
		"authentik_ldap_operations_total",
		metric.WithDescription("Total number of LDAP operations"),
	)
	if err != nil {
		return fmt.Errorf("failed to create LDAP operation counter: %w", err)
	}

	p.ldapOpDuration, err = meter.Float64Histogram(
		"authentik_ldap_operation_duration_seconds",
		metric.WithDescription("Duration of LDAP operations"),
		metric.WithUnit("s"),
	)
	if err != nil {
		return fmt.Errorf("failed to create LDAP operation duration histogram: %w", err)
	}

	return nil
}

// GetTracer returns a tracer for the given name
func (p *Provider) GetTracer(name string) trace.Tracer {
	if !p.initialized {
		return trace.NewNoopTracerProvider().Tracer(name)
	}
	return otel.Tracer(name)
}

// GetMeter returns a meter for the given name
func (p *Provider) GetMeter(name string) metric.Meter {
	if !p.initialized {
		return noop.NewMeterProvider().Meter(name)
	}
	return otel.Meter(name)
}

// RecordRequest records HTTP request metrics
func (p *Provider) RecordRequest(ctx context.Context, method, status string, duration float64) {
	if !p.initialized || p.requestCounter == nil {
		return
	}

	attrs := metric.WithAttributes(
		attribute.String("method", method),
		attribute.String("status", status),
	)

	p.requestCounter.Add(ctx, 1, attrs)
	if p.requestDuration != nil {
		p.requestDuration.Record(ctx, duration, attrs)
	}
}

// RecordAuth records authentication metrics
func (p *Provider) RecordAuth(ctx context.Context, method string, success bool) {
	if !p.initialized || p.authCounter == nil {
		return
	}

	attrs := metric.WithAttributes(
		attribute.String("method", method),
		attribute.Bool("success", success),
	)

	p.authCounter.Add(ctx, 1, attrs)
}

// RecordLDAPOperation records LDAP operation metrics
func (p *Provider) RecordLDAPOperation(ctx context.Context, operation string, duration float64, success bool) {
	if !p.initialized || p.ldapOpCounter == nil {
		return
	}

	attrs := metric.WithAttributes(
		attribute.String("operation", operation),
		attribute.Bool("success", success),
	)

	p.ldapOpCounter.Add(ctx, 1, attrs)
	if p.ldapOpDuration != nil {
		p.ldapOpDuration.Record(ctx, duration, attrs)
	}
}

// Shutdown gracefully shuts down the telemetry provider
func (p *Provider) Shutdown(ctx context.Context) error {
	if !p.initialized {
		return nil
	}

	var err error
	if p.tracerProvider != nil {
		if shutdownErr := p.tracerProvider.Shutdown(ctx); shutdownErr != nil {
			err = fmt.Errorf("failed to shutdown tracer provider: %w", shutdownErr)
		}
	}

	if p.meterProvider != nil {
		if shutdownErr := p.meterProvider.Shutdown(ctx); shutdownErr != nil {
			if err != nil {
				err = fmt.Errorf("%w; failed to shutdown meter provider: %w", err, shutdownErr)
			} else {
				err = fmt.Errorf("failed to shutdown meter provider: %w", shutdownErr)
			}
		}
	}

	p.initialized = false
	return err
}

// getServiceName returns the configured service name or a default
func (p *Provider) getServiceName(serviceName string) string {
	if p.config.Telemetry.OTLP.ServiceName != "" {
		return p.config.Telemetry.OTLP.ServiceName
	}
	if serviceName != "" {
		return serviceName
	}
	return "authentik.outpost"
}

// getServiceVersion returns the configured service version or a default
func (p *Provider) getServiceVersion() string {
	if p.config.Telemetry.OTLP.ServiceVersion != "" {
		return p.config.Telemetry.OTLP.ServiceVersion
	}
	return "unknown"
}

File: authentik/lib/telemetry/provider.py
"""OpenTelemetry provider implementation"""

import logging
import re

from authentik.lib.config import CONFIG

logger = logging.getLogger(__name__)

try:
    from opentelemetry import metrics, trace
    from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
    from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
    from opentelemetry.exporter.otlp.proto.http.metric_exporter import (
        OTLPMetricExporter as HTTPMetricExporter,
    )
    from opentelemetry.exporter.otlp.proto.http.trace_exporter import (
        OTLPSpanExporter as HTTPSpanExporter,
    )
    from opentelemetry.instrumentation.django import DjangoInstrumentor
    from opentelemetry.instrumentation.psycopg2 import Psycopg2Instrumentor
    from opentelemetry.instrumentation.redis import RedisInstrumentor
    from opentelemetry.instrumentation.requests import RequestsInstrumentor
    from opentelemetry.sdk.metrics import MeterProvider
    from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
    from opentelemetry.sdk.resources import SERVICE_NAME, SERVICE_VERSION, Resource
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor

    OPENTELEMETRY_AVAILABLE = True
except ImportError:
    logger.warning("OpenTelemetry packages not available. Telemetry disabled.")
    OPENTELEMETRY_AVAILABLE = False

    # Mock classes for when OpenTelemetry is not available
    class TracerProvider:
        def get_tracer(self, *args, **kwargs):
            return NoOpTracer()

    class MeterProvider:
        def get_meter(self, *args, **kwargs):
            return NoOpMeter()

    class NoOpTracer:
        def start_as_current_span(self, *args, **kwargs):
            return NoOpSpan()

        def start_span(self, *args, **kwargs):
            return NoOpSpan()

    class NoOpMeter:
        def create_counter(self, *args, **kwargs):
            return NoOpCounter()

        def create_histogram(self, *args, **kwargs):
            return NoOpHistogram()

    class NoOpSpan:
        def __enter__(self):
            return self

        def __exit__(self, *args):
            pass

        def set_attribute(self, *args, **kwargs):
            pass

        def set_status(self, *args, **kwargs):
            pass

        def record_exception(self, *args, **kwargs):
            pass

    class NoOpCounter:
        def add(self, *args, **kwargs):
            pass

    class NoOpHistogram:
        def record(self, *args, **kwargs):
            pass


class AdaptiveSampler:
    """Custom sampler that excludes health checks and samples errors at higher rates"""

    def __init__(self, default_rate: float = 0.01, error_rate: float = 1.0):
        self.default_rate = default_rate
        self.error_rate = error_rate
        self.health_check_patterns = [
            re.compile(r"/health/?$"),
            re.compile(r"/metrics/?$"),
            re.compile(r"/readiness/?$"),
            re.compile(r"/liveness/?$"),
        ]

    def should_sample(self, url_path: str, is_error: bool = False) -> bool:
        """Determine if a request should be sampled"""
        # Never sample health check endpoints
        for pattern in self.health_check_patterns:
            if pattern.search(url_path):
                return False

        # Sample errors at higher rate
        if is_error:
            return True  # Always sample errors for now, could use error_rate

        # Default sampling logic would go here
        # For now, return True to sample everything except health checks
        return True


class TelemetryProvider:
    """Central provider for OpenTelemetry telemetry configuration"""

    def __init__(self):
        self._initialized = False
        self._tracer_provider: TracerProvider | None = None
        self._meter_provider: MeterProvider | None = None
        self._tracer = None
        self._meter = None
        self._sampler = AdaptiveSampler()

        # Metrics
        self._flow_duration_histogram = None
        self._policy_duration_histogram = None
        self._auth_counter = None
        self._flow_counter = None
        self._policy_counter = None

    def is_enabled(self) -> bool:
        """Check if telemetry is enabled"""
        return CONFIG.get_bool("telemetry.otlp.enabled", False)

    def initialize(self) -> bool:
        """Initialize OpenTelemetry with OTLP configuration"""
        if self._initialized:
            return True

        if not self.is_enabled():
            logger.debug("OTLP telemetry disabled")
            return False

        if not OPENTELEMETRY_AVAILABLE:
            logger.warning("OpenTelemetry packages not available")
            return False

        try:
            # Configure resource attributes
            resource_attributes = self._get_resource_attributes()
            resource = Resource.create(resource_attributes)

            # Initialize tracing
            self._setup_tracing(resource)

            # Initialize metrics
            self._setup_metrics(resource)

            # Initialize instrumentations
            self._setup_instrumentations()

            self._initialized = True
            logger.info("OpenTelemetry telemetry initialized successfully")
            return True

        except Exception as exc:
            logger.error("Failed to initialize OpenTelemetry telemetry", exc_info=exc)
            return False

    def _get_resource_attributes(self) -> dict:
        """Get resource attributes from configuration"""
        attributes = {
            SERVICE_NAME: CONFIG.get("telemetry.otlp.service_name", "authentik"),
            SERVICE_VERSION: CONFIG.get("telemetry.otlp.service_version", "unknown"),
        }

        # Add custom resource attributes
        custom_attrs = CONFIG.get("telemetry.otlp.resource_attributes", {})
        if isinstance(custom_attrs, dict):
            attributes.update(custom_attrs)

        return attributes

    def _setup_tracing(self, resource: "Resource"):
        """Setup tracing configuration"""
        self._tracer_provider = TracerProvider(resource=resource)

        # Configure OTLP exporter
        endpoint = CONFIG.get("telemetry.otlp.endpoint", "")
        protocol = CONFIG.get("telemetry.otlp.protocol", "grpc")
        headers = CONFIG.get("telemetry.otlp.headers", {})

        if endpoint:
            if protocol.lower() == "http":
                span_exporter = HTTPSpanExporter(
                    endpoint=f"{endpoint}/v1/traces",
                    headers=headers,
                )
            else:  # Default to gRPC
                span_exporter = OTLPSpanExporter(
                    endpoint=endpoint,
                    headers=headers,
                )

            span_processor = BatchSpanProcessor(span_exporter)
            self._tracer_provider.add_span_processor(span_processor)

        # Set global tracer provider
        trace.set_tracer_provider(self._tracer_provider)
        self._tracer = self._tracer_provider.get_tracer("authentik")

    def _setup_metrics(self, resource: "Resource"):
        """Setup metrics configuration"""
        endpoint = CONFIG.get("telemetry.otlp.endpoint", "")
        protocol = CONFIG.get("telemetry.otlp.protocol", "grpc")
        headers = CONFIG.get("telemetry.otlp.headers", {})

        if endpoint:
            if protocol.lower() == "http":
                metric_exporter = HTTPMetricExporter(
                    endpoint=f"{endpoint}/v1/metrics",
                    headers=headers,
                )
            else:  # Default to gRPC
                metric_exporter = OTLPMetricExporter(
                    endpoint=endpoint,
                    headers=headers,
                )

            metric_reader = PeriodicExportingMetricReader(
                exporter=metric_exporter,
                export_interval_millis=60000,  # 1 minute
            )

            self._meter_provider = MeterProvider(
                resource=resource,
                metric_readers=[metric_reader],
            )
        else:
            self._meter_provider = MeterProvider(resource=resource)

        # Set global meter provider
        metrics.set_meter_provider(self._meter_provider)
        self._meter = self._meter_provider.get_meter("authentik")

        # Create metrics
        self._create_metrics()

    def _create_metrics(self):
        """Create application metrics"""
        if not self._meter:
            return

        self._flow_duration_histogram = self._meter.create_histogram(
            name="authentik_flow_execution_duration_seconds",
            description="Duration of flow executions",
            unit="s",
        )

        self._policy_duration_histogram = self._meter.create_histogram(
            name="authentik_policy_evaluation_duration_seconds",
            description="Duration of policy evaluations",
            unit="s",
        )

        self._auth_counter = self._meter.create_counter(
            name="authentik_authentication_total",
            description="Total number of authentication attempts",
        )

        self._flow_counter = self._meter.create_counter(
            name="authentik_flow_execution_total", description="Total number of flow executions"
        )

        self._policy_counter = self._meter.create_counter(
            name="authentik_policy_evaluation_total",
            description="Total number of policy evaluations",
        )

    def _setup_instrumentations(self):
        """Setup automatic instrumentations"""
        try:
            # Instrument Django
            DjangoInstrumentor().instrument()

            # Instrument requests
            RequestsInstrumentor().instrument()

            # Instrument PostgreSQL
            Psycopg2Instrumentor().instrument()

            # Instrument Redis
            RedisInstrumentor().instrument()

            logger.debug("OpenTelemetry instrumentations initialized")
        except Exception as exc:
            logger.warning("Failed to initialize some instrumentations", exc_info=exc)

    def get_tracer(self, name: str = "authentik"):
        """Get a tracer instance"""
        if self._tracer_provider:
            return self._tracer_provider.get_tracer(name)
        return None

    def get_meter(self, name: str = "authentik"):
        """Get a meter instance"""
        if self._meter_provider:
            return self._meter_provider.get_meter(name)
        return None

    def record_flow_execution(self, flow_name: str, duration: float, success: bool):
        """Record flow execution metrics"""
        if not self._flow_duration_histogram or not self._flow_counter:
            return

        attributes = {
            "flow_name": flow_name,
            "success": success,
        }

        self._flow_duration_histogram.record(duration, attributes)
        self._flow_counter.add(1, attributes)

    def record_policy_evaluation(self, policy_name: str, duration: float, result: bool):
        """Record policy evaluation metrics"""
        if not self._policy_duration_histogram or not self._policy_counter:
            return

        attributes = {
            "policy_name": policy_name,
            "result": result,
        }

        self._policy_duration_histogram.record(duration, attributes)
        self._policy_counter.add(1, attributes)

    def record_authentication(self, method: str, success: bool, duration: float):
        """Record authentication metrics"""
        if not self._auth_counter:
            return

        attributes = {
            "method": method,
            "success": success,
        }

        self._auth_counter.add(1, attributes)
